Add SymPy to PSyclone

hiker opened the pull request:
Closes #533.

hiker commented:
Ready for a first review. One thing I am not sure of is if we should add SymPy docs to the UserGuide, atm I only added it to the DevGuide (including details why structure types work).

arporter reviewed (changes requested):
Great stuff Joerg. It's really nice to see so many TODOs and xfails being removed.
All tests and examples are fine with compilation.
The dev. guide still contains a reference to `math_equal` (in transformations.rst).
There are also several mentions in the code/tests:

    PSyclone$ find . -name \*.py -exec grep -H math_equal {} \;
    ./src/psyclone/tests/psyad/transformations/test_assignment_trans.py:@pytest.mark.xfail(reason="issue #1332 literal math_equal() does "
    ./src/psyclone/tests/core/symbolic_maths_test.py:def test_symbolic_math_equal(fortran_reader, expressions):
    ./src/psyclone/tests/core/symbolic_maths_test.py:def test_symbolic_math_equal_structures(fortran_reader, expressions):
    ./src/psyclone/psyir/transformations/arrayrange2loop_trans.py:            # here, or at least something similar to the math_equal
    ./src/psyclone/psyir/transformations/arrayrange2loop_trans.py:            # here, or at least something similar to the math_equal
    ./src/psyclone/domain/common/algorithm/psyir.py:                    # TODO #753 use a better check for equivalence (math_equal)

I'm unsure whether we should say that "1.0 == 1" or whether it should be !=. One for us to debate. Apart from that, it's very minor things.
All changes are pylint clean.
I'll check the coverage in detail next time.

hiker commented on src/psyclone/tests/core/symbolic_maths_test.py:
```diff
@@ -0,0 +1,245 @@
+                                         ("1.23E5", "123000"),
+                                         ("1.23D5", "123000"),
+                                         ("1.0E+3", "1000"),
+                                         ("1.0", "1"),
```
FWIW:
```
    if (1 == 1.0d0) then
        print *,"EQUAL"
    endif
```
prints equal for gfortran and intel. I am pretty sure the standard specifies that the 'smaller' data type gets converted to the 'larger' (i.e. the integer becomes double - see http://rsusu1.rnd.runnet.ru/develop/fortran/prof77/node108.html - not the standard, but first useful google result) - and as such SymPy should report the same.

hiker replied:
I still think this is the right way, since I assume this is what a compiler would do if it evaluates constant expressions. There is a slight potential problem if you have a huge int, which gets converted to a real, and loses precision. SymPy does not lose precision. so this is the only case that I would see a problem - but I think that comparing 'small' numbers is what is happening in 99% of all cases anyway.

hiker commented on src/psyclone/psyir/backend/fortran.py:
```diff
@@ -405,6 +338,72 @@ def __init__(self, skip_nodes=False, indent_string="  ",
+            if mapping_key not in reverse_dict:
+                reverse_dict[mapping_key] = mapping_value.upper()
+
+    def is_intrinsic(self, operator):
```
IMHO the original design is very bad:
1) It pollutes the namespace by adding some random global functions (what happens if another module defines a function with the same name? Might then the result depend on the order in which imports happen???). That's why I moved it into the class (originally as static function).
2) But even then, the bad design is that the instance of the Fortran writer depends on global variables from a different module. So you can create an instance, use the function, and it works. But them 'something' modifies a totally unrelated variable in a different module, and suddenly this class could break. Seriously ... what is it with all this global stuff everywhere?  :)   IMHO this function should copy the values when the instance is created, and then use its own copy, to be truly self contained. But since this was not the focus of this PR, and I did not want to interfere with existing code, I did not fix these design problems.
3) The SymPy writer does it properly - creating the data in the constructor, and then be self-contained. But that of course requires `self`. And I think it's better to be consistent with the various function, so I added the `self` in the Fortran base class, (and added a directive to ignore the pylint warning).

I am happy to also fix this  - I am just trying to be good and keep PRs small 😊

arporter replied:
Those sound like good points. Perhaps we should have a separate issue to remove the global-variable dependence from FortranWriter.

hiker commented on src/psyclone/psyir/backend/fortran.py:
```diff
@@ -405,6 +338,72 @@ def __init__(self, skip_nodes=False, indent_string="  ",
+
+        :returns: true if the supplied Fortran operator is a Fortran \
+            intrinsic and false otherwise.
+
```
Done.

hiker commented on src/psyclone/tests/core/symbolic_maths_test.py:
```diff
@@ -0,0 +1,245 @@
+    that sympy can be imported.'''
+
+    sym_maths = SymbolicMaths.get()
+    assert sym_maths is not None
```
Done.

hiker commented on src/psyclone/tests/psyad/transformations/test_assignment_trans.py:
```diff
@@ -646,7 +644,6 @@ def test_same_indices_ordering2(tmpdir):
     check_adjoint(tl_fortran, active_variables, ad_fortran, tmpdir)
 
 
-@pytest.mark.xfail(reason="issue #1332 math_equal thinks a%b(i) equals a%c(i)")
```
Party party

hiker commented on doc/user_guide/getting_going.rst:
```diff
@@ -327,7 +327,41 @@ you can instruct ``pip`` to do a user-local install:
+
+   > pip install --user sympy
+
+Alternatively, you could follow on the `SymPy web page
```
Done.

hiker commented on src/psyclone/core/symbolic_maths.py:
```diff
@@ -0,0 +1,110 @@
+    # -------------------------------------------------------------------------
+    @staticmethod
+    def get():
+        '''Static function that if necessary creates and returns the singleton
```
Done.

hiker commented on src/psyclone/core/symbolic_maths.py:
```diff
@@ -0,0 +1,110 @@
+
+    # -------------------------------------------------------------------------
+    def __init__(self):
+
```
Done.

hiker commented on src/psyclone/core/symbolic_maths.py:
```diff
@@ -0,0 +1,110 @@
+
+        str_exp1 = parse_expr(self._writer(exp1))
+        str_exp2 = parse_expr(self._writer(exp2))
+        return simplify(str_exp1 == str_exp2)
```
I've added a comment to explain this (in short, there are several ways that sympy evaluates expression, and 'simplify' is a mixture/order of operations that works well - it's not clearly specified exactly that it does, and it warns that it might change from one version to the other). In my tests it does exactly what I wanted, without the need to find our own best way of applying operations.

arporter replied:
OK, thanks for explaining.

hiker commented on src/psyclone/psyir/backend/sympy_writer.py:
```diff
@@ -0,0 +1,144 @@
+        self._intrinsic = set()
+        self._op_to_str = {}
+
+        # Get the mapping of special operators/functions to the
```
Yes, that's better.

hiker commented on src/psyclone/psyir/backend/sympy_writer.py:
```diff
@@ -0,0 +1,144 @@
+        '''This method is called when a Literal instance is found in the PSyIR
+        tree. For SymPy we need to handle booleans (which are expected to
+        be capitalised: True). Real values work by just ignoring any precision
+        information (e.g. 2_4, 3.1_wp). Character constants are supported
```
Oops :(

hiker commented on src/psyclone/tests/core/symbolic_maths_test.py:
```diff
@@ -0,0 +1,245 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2019-2021, Science and Technology Facilities Council.
```
Yep.

hiker commented on src/psyclone/tests/core/symbolic_maths_test.py:
```diff
@@ -0,0 +1,245 @@
+# Modified: I. Kavcic, Met Office
+
+
+''' Module containing py.test tests for dependency analysis.'''
```
Done.

hiker commented on src/psyclone/tests/core/symbolic_maths_test.py:
```diff
@@ -0,0 +1,245 @@
+                                         (".false.", ".FALSE."),
+                                         ])
+def test_math_logicals(fortran_reader, expressions):
+    '''Test that the sympy based comparison handles constant numbers,
```
Done.

hiker commented on src/psyclone/tests/core/symbolic_maths_test.py:
```diff
@@ -0,0 +1,245 @@
+    # A dummy program to easily create the PSyIR for the
+    # expressions we need. We just take the RHS of the assignments
+    source = '''program test_prog
+                use some_mod
```
Indeed not.

hiker commented on src/psyclone/tests/core/symbolic_maths_test.py:
```diff
@@ -0,0 +1,245 @@
+                                         ("i+j+k", "i+k+j"),
+                                         ("i+i", "2*i"),
+                                         ("i+j-2*k+3*j-2*i", "-i+4*j-2*k"),
+                                         ("max(1, 2, 3)", "max(1, 2, 3)")
```
I forgot to remove this (that test highlighted the issue of SymPy function names). I've removed this test here, further down in the file are various tests with max (and min), including different order of parameters.

hiker commented on src/psyclone/tests/core/symbolic_maths_test.py:
```diff
@@ -0,0 +1,245 @@
+                                         ])
+def test_symbolic_math_not_equal(fortran_reader, expressions):
+    '''Test that the sympy based comparison handles complex
+    expressions.
```
Done.

hiker commented on src/psyclone/tests/core/symbolic_maths_test.py:
```diff
@@ -0,0 +1,245 @@
+    schedule = psyir.children[0]
+
+    sym_maths = SymbolicMaths.get()
+    # Note we cannot use 'is False', since sym_maths returns an
```
As explained elsewhere, I've changed that (including tests now using `is True` or `is False`).

hiker commented on src/psyclone/tests/core/symbolic_maths_test.py:
```diff
@@ -0,0 +1,245 @@
+                                         ])
+def test_symbolic_math_not_equal_structures(fortran_reader, expressions):
+    '''Test that the sympy based comparison handles complex
+    expressions.
```
Done.

hiker commented on src/psyclone/tests/core/symbolic_maths_test.py:
```diff
@@ -0,0 +1,245 @@
+
+    sym_maths = SymbolicMaths.get()
+
+    # Note we cannot use 'is False', since sym_maths returns an
```
Done.

hiker commented on src/psyclone/tests/core/symbolic_maths_test.py:
```diff
@@ -0,0 +1,245 @@
+                                         ("MOD(i,j)", "mod(2+i-2, j)")
+                                         ])
+def test_symbolic_math_functions_with_constants(fortran_reader, expressions):
+    '''Test how known functions with constant values are handled.
```
Done.

hiker commented on src/psyclone/tests/psyir/backend/sympy_writer_test.py:
```diff
@@ -0,0 +1,143 @@
+def test_sym_writer_character():
+    '''Test that characters are rejected.
+    '''
+
```
Done.

hiker commented on src/psyclone/tests/psyir/backend/sympy_writer_test.py:
```diff
@@ -0,0 +1,143 @@
+    sympy_writer = SymPyWriter()
+    lit = Literal("bla", CHARACTER_TYPE)
+
+    with pytest.raises(InternalError) as err:
```
Good idea - all changed.

hiker commented:
> Great stuff Joerg. It's really nice to see so many TODOs and xfails being removed. All tests and examples are fine with compilation. The dev. guide still contains a reference to `math_equal` (in transformations.rst). There are also several mentions in the code/tests:

Oops, I missed that. Fixed.
> 
> ```
> PSyclone$ find . -name \*.py -exec grep -H math_equal {} \;
> ./src/psyclone/tests/psyad/transformations/test_assignment_trans.py:@pytest.mark.xfail(reason="issue #1332 literal math_equal() does "

OK, I had a look at #1332 and this x-failing test. The reason it is x-failing is that the comparison with the known-good-outcome is just a string comparison, and the output is in the wrong order (but semantically correct). SymPy **can** detect that the expressions are equal (I've added this to #1332). I've just changed the input to be in the correct order, and now the test passes.

If the intention of #1332 is to write a test that compares a program 'more flexible' (i.e. not just a string comparison), that should be possible with sympy,  but is not obvious in #1332. I've removed the x-failing and reference to #1332. IMHO #1332 can then be closed when this is merged.

> ./src/psyclone/psyir/transformations/arrayrange2loop_trans.py:            # here, or at least something similar to the math_equal
> ./src/psyclone/psyir/transformations/arrayrange2loop_trans.py:            # here, or at least something similar to the math_equal
(I think I did a find on `math_equal(` *sigh*). I've fixed that - not just the reference, but added SymPy tests, and it passed all tests. I assume that's ok.

> ./src/psyclone/domain/common/algorithm/psyir.py:                    # TODO #753 use a better check for equivalence (math_equal)

I've fixed that ... but I actually run into a bug: the test case uses a variable called `field`, and `field` is a SymPy function ... and because of this parsing fails (not really obvious, parsing uses eval, and then things go wrong).  I've opened #1523 to dig into this (simply renaming `field` to `field` would atm fix the test).

hiker commented:
Luckily, #1523 seems to have a nice solution without too much effort. 

Ready for next review.

arporter reviewed (changes requested):
I think we're almost there now.
There's still a reference to math_equal in the 'transformations' section of the dev. guide.
I get a test failure when compilation is enabled:

    psyad/transformations/test_assignment_trans.py:110: AssertionError
    ----------------------------- Captured stderr call -----------------------------
    Compiling: gfortran generated-8750288566215.f90 -c
    b'generated-8750288566215.f90:2:6:\n\n    2 |   use field_mod, only : field_type\n      |      1\nFatal Error: Cannot open module file \xe2\x80\x98field_mod.mod\xe2\x80\x99 for reading at (1): No such file or directory\ncompilation terminated

This doesn't happen on master - please could you investigate? Perhaps it was a previously x-failing test?

It looks as though the ArrayRange2LoopTrans.string_compare() method can be removed completely.

Other than that, the examples are fine and the changes are all pylint clean.
The new sympy_writer tests cover the new file completely. Ditto the symbolic_maths tests.

hiker commented on src/psyclone/psyir/transformations/arrayrange2loop_trans.py:
```diff
@@ -218,17 +220,14 @@ def same_range(array1, idx1, array2, idx2):
             # different (although they could potentially be the same).
             return False
-        elif not ArrayRange2LoopTrans.string_compare(range1.stop, range2.stop):
+        elif not sym_maths.equal(range1.stop, range2.stop):
```
Indeed, I have removed it.

hiker commented:
> I think we're almost there now. There's still a reference to math_equal in the 'transformations' section of the dev. guide.

I felt so stupid and coulnd't believe it, I was certain that I grep'ed properly ... but I have a good explanation at least ... when I tried to fix an failing issue, I ran into #1523. So I stashed my fix ... not realising that I also stashed the documentation change. Now I have applied my fix, and changed the x-failing test to properly point to #1523 now.


>  I get a test failure when compilation is enabled:
...
> This doesn't happen on master - please could you investigate? Perhaps it was a previously x-failing test?

Yes, it was indeed previously x-failing. The compilation code is incorrect (it calls the generic compile function, not the LFRic-specific one, which supplies the path to the infrastructure library), so I've fixed this..
But: compilation now still fails ... because the created code is invalid. I've opened #1530 , and fixed up the test for now to be successful when no compilation is requested, and x-fail if we indeed trigger an assertion (.

Assuming that the tests pass, ready for next review.

arporter commented on src/psyclone/tests/psyad/transformations/test_assignment_trans.py:
```diff
@@ -665,7 +666,14 @@ def test_different_structures(tmpdir):
         "  a%atad(n) = a%atad(n) + a%data(n)\n"
         "  a%data(n) = 0.0\n\n")
-    check_adjoint(tl_fortran, active_variables, ad_fortran, tmpdir)
+
```
Thanks for attempting to fix our mess :-) I don't think this test is intended to be specific to the LFRic API. The problem obviously is that it imports `field_type` from `field_mod` and we can't compile that. Please could you revert to just using the generic Compile class and leave the compilation of this xfailing with a TODO that we need to
explicitly define the field_type rather than importing it from a module?

hiker replied:
OK, I've changed #1530 to correctly describe the problem (field_type should be defined, not imported from LFRic), and reverted LFRicBuild to be Compile (and left it as xfailing). Hope I got it correctly now :)

arporter reviewed (changes requested):
Almost, almost there (and it's not really your fault that it's not).
Just a bit of undoing to do and then this can go on (provided compilation etc. is all OK).

arporter reviewed (approved):
Thanks for updating the change.
All tests and examples are OK with compilation.
All docs build fine.
